File: apps/cli/src/commands/update.ts
import * as p from "@clack/prompts";
import pc from "picocolors";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { exec } from "node:child_process";
import { promisify } from "node:util";
import { printBanner } from "../ui/banner.js";
import { ReleaseService } from "../services/release.js";

const execAsync = promisify(exec);
const CLI_VERSION = "0.1.0";

async function pathExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

export async function handleUpdateCommand(): Promise<void> {
  printBanner(CLI_VERSION);

  p.intro(pc.bgCyan(pc.black(" 🔄 Amoeba Project & CLI Updater ")));

  const s = p.spinner();

  // 1. Check for CLI Release / Patch updates
  s.start("Checking for CLI release & patch updates...");
  const releaseService = new ReleaseService();
  const releaseInfo = await releaseService.checkLatestRelease(CLI_VERSION);
  s.stop(pc.green("✔ Release check complete"));

  if (releaseInfo.hasUpdate) {
    const typeColor =
      releaseInfo.updateType === "major"
        ? pc.red
        : releaseInfo.updateType === "minor"
        ? pc.yellow
        : pc.cyan;

    const message = [
      `${pc.bold("New update available:")} ${pc.dim(`v${releaseInfo.currentVersion}`)} → ${typeColor(pc.bold(`v${releaseInfo.latestVersion}`))} (${releaseInfo.updateType} release)`,
      releaseInfo.releaseUrl ? `${pc.bold("Release details:")}      ${pc.blue(releaseInfo.releaseUrl)}` : "",
      releaseInfo.releaseNotes ? `\n${pc.bold("Changelog:")}\n${pc.dim(releaseInfo.releaseNotes.slice(0, 300))}` : "",
      `\n${pc.bold("To update Amoeba CLI run:")}\n  ${pc.cyan("git pull && pnpm --filter amoeba-cli build")}`,
    ]
      .filter(Boolean)
      .join("\n");

    p.note(message, "⚡ CLI Update Available");
  } else {
    p.log.info(pc.dim(`Amoeba CLI is up to date (v${CLI_VERSION})`));
  }

  // 2. Check for local Amoeba project components
  const cwd = process.cwd();
  const apiDir = path.join(cwd, "apps", "api");
  const webDir = path.join(cwd, "apps", "web");
  const desktopDir = path.join(cwd, "apps", "desktop");

  const hasAPI = await pathExists(path.join(apiDir, "go.mod"));
  const hasWeb = await pathExists(path.join(webDir, "package.json"));
  const hasDesktop = await pathExists(path.join(desktopDir, "package.json"));

  if (!hasAPI && !hasWeb && !hasDesktop) {
    p.log.step(
      pc.dim("No local project dependencies to update (not inside an Amoeba workspace directory).")
    );
    p.outro(pc.bold(pc.green("✔ Check finished!")));
    return;
  }

  // 3. Update Go API dependencies
  if (hasAPI) {
    s.start("Updating Go backend dependencies (Fiber v3 & database drivers)...");
    try {
      await execAsync("go get -u ./... && go mod tidy", { cwd: apiDir });
      s.stop(pc.green("✔ Go backend dependencies updated"));
    } catch (err) {
      s.stop(pc.red("✖ Failed to update Go dependencies"));
      p.log.error(err instanceof Error ? err.message : String(err));
    }
  }

  // 4. Update Frontend dependencies
  const frontendDir = hasWeb ? webDir : hasDesktop ? desktopDir : null;
  const frontendName = hasWeb ? "Web (apps/web)" : hasDesktop ? "Desktop (apps/desktop)" : null;

  if (frontendDir && frontendName) {
    s.start(`Updating frontend dependencies in ${frontendName}...`);
    try {
      let updateCmd = "pnpm up --latest";
      try {
        await execAsync("pnpm --version");
      } catch {
        updateCmd = "npm update";
      }

      await execAsync(updateCmd, { cwd: frontendDir });
      s.stop(pc.green(`✔ ${frontendName} dependencies updated`));
    } catch (err) {
      s.stop(pc.red(`✖ Failed to update ${frontendName} dependencies`));
      p.log.error(err instanceof Error ? err.message : String(err));
    }
  }

  p.outro(pc.bold(pc.green("⚡ All Amoeba components up to date!")));
}

File: apps/cli/src/commands/update.rs
use anyhow::Result;
use colored::Colorize;
use indicatif::{ProgressBar, ProgressStyle};
use std::env;
use std::process::Command;

use crate::banner::print_banner;
use crate::utils::fs::path_exists;
use crate::utils::release::check_latest_release;

const CLI_VERSION: &str = "0.2.1";

pub fn handle_update_command() -> Result<()> {
    print_banner(CLI_VERSION);

    println!(
        "{}\n",
        " ⚡ Amoeba Project & CLI Updater ".on_cyan().black().bold()
    );

    let pb = ProgressBar::new_spinner();
    pb.set_style(
        ProgressStyle::default_spinner()
            .tick_chars("⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏ ")
            .template("{spinner:.green} {msg}")?,
    );

    // 1. Release check
    pb.set_message("Checking for latest Amoeba releases on GitHub...");
    let release_info = check_latest_release(CLI_VERSION)?;
    pb.finish_with_message(format!("{}", "✔ Release check complete".green()));

    if release_info.has_update {
        println!(
            "\n{} {}",
            "⚡ New Amoeba Release Detected:".cyan().bold(),
            format!("v{}", release_info.latest_version).green().bold()
        );
        println!("  Current version: {}", format!("v{}", CLI_VERSION).dimmed());
        if let Some(url) = release_info.release_url {
            println!("  Release notes:   {}", url.blue().underline());
        }
    } else {
        println!(
            "\n{} Amoeba CLI is running the latest release ({})",
            "✔".green().bold(),
            format!("v{}", CLI_VERSION).cyan()
        );
    }

    // 2. Local project component checks
    let cwd = env::current_dir()?;
    let api_go = cwd.join("apps/api/go.mod");
    let api_ts = cwd.join("apps/api/package.json");
    let web_pkg = cwd.join("apps/web/package.json");
    let desktop_pkg = cwd.join("apps/desktop/package.json");
    let workspace_pkg = cwd.join("pnpm-workspace.yaml");

    let has_go_api = path_exists(&api_go);
    let has_ts_api = path_exists(&api_ts);
    let has_web = path_exists(&web_pkg);
    let has_desktop = path_exists(&desktop_pkg);
    let has_workspace = path_exists(&workspace_pkg);

    if !has_go_api && !has_ts_api && !has_web && !has_desktop && !has_workspace {
        println!(
            "\n{}",
            "No local Amoeba project detected in current directory.".dimmed()
        );
        println!("{}", "⚡ Update check complete!".green().bold());
        return Ok(());
    }

    println!("\n{}", "Updating project dependencies...".white().bold());

    // Update Go API
    if has_go_api {
        let api_dir = cwd.join("apps/api");
        println!("  {} Updating Go Fiber & database drivers...", "•".cyan());
        let _ = Command::new("go")
            .args(["get", "-u", "./..."])
            .current_dir(&api_dir)
            .status();
        let _ = Command::new("go")
            .args(["mod", "tidy"])
            .current_dir(&api_dir)
            .status();
        println!("    {}", "✔ Go modules updated and tidied".green());
    }

    // Update Node/TS workspace or apps
    if has_workspace {
        println!("  {} Updating monorepo workspace dependencies...", "•".cyan());
        let _ = Command::new("pnpm")
            .args(["update"])
            .current_dir(&cwd)
            .status();
        println!("    {}", "✔ Monorepo dependencies updated".green());
    } else {
        if has_ts_api {
            println!("  {} Updating Express API dependencies...", "•".cyan());
            let _ = Command::new("pnpm")
                .args(["update"])
                .current_dir(cwd.join("apps/api"))
                .status();
        }
        if has_web {
            println!("  {} Updating Web frontend dependencies...", "•".cyan());
            let _ = Command::new("pnpm")
                .args(["update"])
                .current_dir(cwd.join("apps/web"))
                .status();
        }
        if has_desktop {
            println!("  {} Updating Desktop frontend dependencies...", "•".cyan());
            let _ = Command::new("pnpm")
                .args(["update"])
                .current_dir(cwd.join("apps/desktop"))
                .status();
        }
    }

    println!("\n{}", "⚡ Project update completed successfully!".green().bold());
    Ok(())
}
